src/management/list_files.rs

```diff
@@ -0,0 +1,92 @@
+use anyhow::{bail, Result};
+use async_trait::async_trait;
+use reqwest::StatusCode;
+
+use crate::{client::FILES_ENDPOINT, upload::types::Response, ErrorResponse, ImageKit};
+
+/// Options for list/search files.
+///
+/// Refer: https://docs.imagekit.io/api-reference/media-api/list-and-search-files
+#[derive(Default)]
+pub struct Options {
+    search_query: Option<String>,
```

Comment:
Do you think we could use a method instead of raw string?

One possibility is using a `enum` with the Query params.

```rust
enum Search {
  Name(String)
}
```

Then we could have `search_query` to be: `HashSet<Search>`.

So, whenever we want to add a Query param we can push to the `HashSet`.

`search_query.push(Search::Name("mypicture".into())`

Then when we want to actually perform the search or build the URL, we can iteare over the `HashSet<Search>` and build the URL.

Let me know your thoughts!

Reply:
I'm not sure I completely got your thought.

Based on what I can guess, it sounds like a good thought for simple use-cases. However, what I feel is we would eventually need some sort of a 'builder' to help constructing this raw query string, while still retaining the ability to use 'the raw query string' (especially, for those who are comfortable with it (like those coming from other language binding, or , who already have the raw query string handy, or, those who are comfortable with raw rest API Paramus))

btw, a simple hashset (or even a custom struct) won't cover some of the advanced usecases, as the query string could have nesting, with AND's, OR's, operators and brackets mixed at user's convenience. So, it is actually kind of a 'tree' mixed with operator's, 'and' and 'or'.  So, I feel we need some sort of a builder to do this stuff, if we really need a typed search query construction. 

A basic example from the documentation goes like this.. "(size < "1mb" AND width > 500) OR (tags IN ["summer-sale","banner"])"

a rough pseudo-code of what I think could be..

query_builder.add(Search::NameEquals("mypicture".into()).or(Search::NameLike("mypicture2".into()).and.... (and so on). And, I haven't tried simulating a more complex ombinations, along with 'operators'. Operators' for example, supports the following just for name '=, :, IN, NOT =, NOT IN). And the operators' varies per each parameter.

So, this is defintely a more complex stuff to provide a good typed behavior. I'm afraid we may end up similar to a rule engine like https://docs.rs/rule/latest/rule/, to really cover the advanced use-cases and combinations, or may be like a SQL like processor :)

So, my thought process is to deal with this typed query_string construction complexity probably at an additional iteration, if at all "possible".

Saying this, I'm not sure if I completely got your thoughts as you envisioned. My interpretation could be wrong also. Is there a discord channel or other chat options, where we could communicate more interactively? If not, we can continue here itself :)

Reply:
> I'm not sure I completely got your thought.
> 
> Based on what I can guess, it sounds like a good thought for simple use-cases. However, what I feel is we would eventually need some sort of a 'builder' to help constructing this raw query string, while still retaining the ability to use 'the raw query string' (especially, for those who are comfortable with it (like those coming from other language binding, or , who already have the raw query string handy, or, those who are comfortable with raw rest API Paramus))
> 
> btw, a simple hashset (or even a custom struct) won't cover some of the advanced usecases, as the query string could have nesting, with AND's, OR's, operators and brackets mixed at user's convenience. So, it is actually kind of a 'tree' mixed with operator's, 'and' and 'or'. So, I feel we need some sort of a builder to do this stuff, if we really need a typed search query construction.
> 
> A basic example from the documentation goes like this.. "(size < "1mb" AND width > 500) OR (tags IN ["summer-sale","banner"])"
> 
> a rough pseudo-code of what I think could be..
> 
> query_builder.add(Search::NameEquals("mypicture".into()).or(Search::NameLike("mypicture2".into()).and.... (and so on). And, I haven't tried simulating a more complex ombinations, along with 'operators'. Operators' for example, supports the following just for name '=, :, IN, NOT =, NOT IN). And the operators' varies per each parameter.
> 
> So, this is defintely a more complex stuff to provide a good typed behavior. I'm afraid we may end up similar to a rule engine like https://docs.rs/rule/latest/rule/, to really cover the advanced use-cases and combinations, or may be like a SQL builder kind of stuff :)
> 
> So, my thought process is to deal with this typed query_string construction complexity probably at an additional iteration, if at all "possible".
> 
> Saying this, I'm not sure if I completely got your thoughts as you envisioned. My interpretation could be wrong also. Is there a discord channel or other chat options, where we could communicate more interactively? If not, we can continue here itself :)

@EstebanBorai btw, I got some thought on a 'middle ground' on this approach. Will update the code shortly and see how that approach looks for you. Will comment here once I make some basic structure in place.

Reply:
Please hold on with the merge in the meanwhile @EstebanBorai